src/gba/vfame.c

```diff
@@ -0,0 +1,324 @@
+/* Copyright (c) 2016 taizou
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "vfame.h"
+#include "memory.h"
+
+static const char ADDRESS_REORDERING_1[16] = { 15, 14, 9, 1, 8, 10, 7, 3, 5, 11, 4, 0, 13, 12, 2, 6 };
+static const char ADDRESS_REORDERING_2[16] = { 15, 7, 13, 5, 11, 6, 0, 9, 12, 2, 10, 14, 3, 1, 8, 4 };
+static const char ADDRESS_REORDERING_3[16] = { 15, 0, 3, 12, 2, 4, 14, 13, 1, 8, 6, 7, 9, 5, 11, 10 };
+static const char ADDRESS_REORDERING_GEORGE_1[16] = { 15, 7, 13, 1, 11, 10, 14, 9, 12, 2, 4, 0, 3, 5, 2, 6 };
+static const char ADDRESS_REORDERING_GEORGE_2[16] = { 15, 14, 3, 12, 8, 4, 0, 13, 5, 11, 6, 7, 9, 1, 2, 10 };
+static const char ADDRESS_REORDERING_GEORGE_3[16] = { 15, 0, 9, 5, 2, 6, 7, 3, 1, 8, 10, 14, 13, 12, 11, 4 };
+static const char VALUE_REORDERING_1[8] = { 5, 4, 3, 2, 1, 0, 7, 6 };
+static const char VALUE_REORDERING_2[8] = { 3, 2, 1, 0, 7, 6, 5, 4 };
+static const char VALUE_REORDERING_3[8] = { 1, 0, 7, 6, 5, 4, 3, 2 };
+static const char VALUE_REORDERING_GEORGE_1[8] = { 3, 0, 7, 2, 1, 4, 5, 6 };
+static const char VALUE_REORDERING_GEORGE_2[8] = { 1, 4, 3, 0, 5, 6, 7, 2 };
+static const char VALUE_REORDERING_GEORGE_3[8] = { 5, 2, 1, 6, 7, 0, 3, 4 };
+
+static const int8_t MODE_CHANGE_START_SEQUENCE[5] = { 0x99, 0x02, 0x05, 0x02, 0x03 };
+static const int8_t MODE_CHANGE_END_SEQUENCE[5] = { 0x99, 0x03, 0x62, 0x02, 0x56 };
+
+// A portion of the initialisation routine that gets copied into RAM - Always seems to be present at 0x15C in VFame game ROM
+static const char INIT_SEQUENCE[16] = { 0xB4, 0x00, 0x9F, 0xE5, 0x99, 0x10, 0xA0, 0xE3, 0x00, 0x10, 0xC0, 0xE5, 0xAC, 0x00, 0x9F, 0xE5 };
+
+static bool _isInMirroredArea(uint32_t address, size_t romSize);
+static uint32_t _getPatternValue(uint32_t addr);
+static uint32_t _patternRightShift1(uint32_t addr);
+static uint32_t _patternRightShift2(uint32_t addr);
+static int8_t _modifySramValue(enum GBAVFameCartType type, int8_t value, int mode);
+static uint32_t _modifySramAddress(enum GBAVFameCartType type, uint32_t address, int mode);
+static int _reorderBits(uint32_t value, const char* reordering, int reorderLength);
+
+void GBAVFameInit(struct GBAVFameCart* cart) {
+	cart->cartType = VFAME_NO;
+	cart->sramMode = -1;
+	cart->romMode = -1;
+	cart->acceptingModeChange = false;
+}
+
+void GBAVFameDetect(struct GBAVFameCart* cart, uint32_t* rom, size_t romSize) {
+	cart->cartType = VFAME_NO;
+
+	// The initialisation code is also present & run in the dumps of Digimon Ruby & Sapphire from hacked/deprotected reprint carts,
+	// which would break if run in "proper" VFame mode so we need to exclude those..
+	if (romSize == 0x2000000) { // the deprotected dumps are 32MB but no real VF games are this size
+		return;
+	}
+
+	if (memcmp(INIT_SEQUENCE, &((char*)rom)[0x15C], 16) == 0) {
```

Comment:
`memcmp(INIT_SEQUENCE, &rom[0x57], sizeof(INIT_SEQUENCE)) == 0` is a bit cleaner.